File: engine/sweep.ts
/**
 * sweep.ts — Suspension Travel Sweep Engine
 *
 * Computes MR(u), WR(u), AS%(u), and trail(u) over the full rear
 * suspension travel range by simulating swingarm kinematics at each
 * 1 mm step of wheel travel.
 *
 * Two rear suspension topologies are supported:
 *
 *   'direct'  — shock connects swingarm directly to frame (monoshock)
 *   'fourbar' — Pro-Link / Uni-Trak: rocker + pushrod between swingarm
 *               and shock.  One nonlinear equation (loop closure) solved
 *               by Newton-Raphson at each step.
 *
 * COORDINATE SYSTEM
 *   Origin = front tyre contact patch, level ground.
 *   +X = rearward (positive toward rear axle).
 *   +Y = upward.
 *   All lengths in mm.
 *
 * REFERENCES
 *   Foale (2006) Ch. 6 — Motion ratio, wheel rate
 *   Foale (2006) Ch. 11 — Anti-squat, instant centre method
 *   Cossalter (2006) Ch. 5 — Suspension kinematics
 */

import { GeometryParams, SuspensionParams, ChainParams, SweepParams, SweepPoint, SweepResults } from './types';

const DEG2RAD = Math.PI / 180;
const RAD2DEG = 180 / Math.PI;

// ─── Swingarm kinematics ──────────────────────────────────────────────────────

/**
 * Compute swingarm angle at wheel travel u.
 *
 * The rear axle rides on a circular arc of radius L_sa centred on the
 * swingarm pivot.  Wheel travel u (positive = bump / compression) raises
 * the rear axle: H_ra(u) = H_ra_static + u.
 *
 * θ_sa(u) = arcsin((H_ra(u) − H_sp) / L_sa)
 */
function swingarmAngleAtTravel(
  u_mm: number,
  H_ra_static: number,
  H_sp: number,
  L_sa: number,
): number {
  const H_ra_u = H_ra_static + u_mm;
  const sinTheta = (H_ra_u - H_sp) / L_sa;
  // clamp to [-1, 1] to avoid asin domain errors at geometry limits
  return Math.asin(Math.max(-1, Math.min(1, sinTheta)));
}

// ─── Shock length helpers ─────────────────────────────────────────────────────

/**
 * World-frame position of the shock attachment point on the swingarm.
 *
 * The attachment is at distance d from the pivot, at angle (θ_sa + φ_arm)
 * measured from the horizontal.
 */
function shockAttachWorld(
  X_sp: number, H_sp: number,
  theta_sa: number,
  d: number, phi_arm_rad: number,
): [number, number] {
  const angle = theta_sa + phi_arm_rad;
  return [X_sp + d * Math.cos(angle), H_sp + d * Math.sin(angle)];
}

/** Euclidean distance between two 2-D points. */
function dist2(ax: number, ay: number, bx: number, by: number): number {
  const dx = bx - ax, dy = by - ay;
  return Math.sqrt(dx * dx + dy * dy);
}

// ─── Direct monoshock ─────────────────────────────────────────────────────────

function shockLengthDirect(
  u_mm: number,
  gp: GeometryParams,
  sp: SweepParams,
): number {
  const theta = swingarmAngleAtTravel(
    u_mm, gp.rearAxleHeight, gp.swingarmPivotHeight, gp.swingarmLength,
  );
  const [ax, ay] = shockAttachWorld(
    gp.swingarmPivotX, gp.swingarmPivotHeight,
    theta, sp.shockArmLength, sp.shockArmAngle * DEG2RAD,
  );
  return dist2(ax, ay, sp.shockTopX, sp.shockTopY);
}

// ─── 4-bar (Pro-Link / Uni-Trak) ─────────────────────────────────────────────

/**
 * Solve the single-DOF 4-bar loop closure for rocker angle θ_r.
 *
 * Given the swingarm attachment S(u), rocker pivot R, rocker length
 * L_rock and pushrod length L_push, find θ_r such that:
 *
 *   |Q − S|² = L_push²   where Q = R + L_rock·[cos θ_r, sin θ_r]
 *
 * Expanded: f(θ_r) = (R_x + L_rock·cos θ_r − S_x)² +
 *                    (R_y + L_rock·sin θ_r − S_y)² − L_push² = 0
 *
 * Solved by Newton-Raphson starting from the previous step's angle.
 */
function solveRockerAngle(
  S: [number, number],
  R: [number, number],
  L_rock: number,
  L_push: number,
  theta_r0: number,
): number {
  let theta = theta_r0;
  for (let iter = 0; iter < 50; iter++) {
    const Qx = R[0] + L_rock * Math.cos(theta);
    const Qy = R[1] + L_rock * Math.sin(theta);
    const dx = Qx - S[0], dy = Qy - S[1];
    const f  = dx * dx + dy * dy - L_push * L_push;
    const df = 2 * (-L_rock * Math.sin(theta) * dx + L_rock * Math.cos(theta) * dy);
    if (Math.abs(df) < 1e-12) break;
    const dtheta = -f / df;
    theta += dtheta;
    if (Math.abs(dtheta) < 1e-9) break;
  }
  return theta;
}

function shockLengthFourBar(
  u_mm: number,
  gp: GeometryParams,
  sp: SweepParams,
  theta_r0: number,
): { shockLen: number; theta_r: number } {
  const fb = sp.fourBar!;
  const theta_sa = swingarmAngleAtTravel(
    u_mm, gp.rearAxleHeight, gp.swingarmPivotHeight, gp.swingarmLength,
  );
  // Swingarm attachment (S)
  const [Sx, Sy] = shockAttachWorld(
    gp.swingarmPivotX, gp.swingarmPivotHeight,
    theta_sa, sp.shockArmLength, sp.shockArmAngle * DEG2RAD,
  );
  // Rocker pivot (R) — fixed to frame
  const R: [number, number] = [fb.rockerPivotX, fb.rockerPivotY];

  // Solve for rocker angle
  const theta_r = solveRockerAngle(
    [Sx, Sy], R, fb.rockerLength, fb.pushrodLength, theta_r0,
  );

  // Rocker tip Q
  const Qx = R[0] + fb.rockerLength * Math.cos(theta_r);
  const Qy = R[1] + fb.rockerLength * Math.sin(theta_r);

  // Shock length = |Q − shock_top|
  const shockLen = dist2(Qx, Qy, sp.shockTopX, sp.shockTopY);
  return { shockLen, theta_r };
}

// ─── Anti-squat at travel position ────────────────────────────────────────────

/**
 * Compute anti-squat % for a given swingarm angle using the Foale
 * graphical / instant-centre method.
 *
 * Reproduces the logic in antiSquat.ts but as a pure function so the
 * sweep engine doesn't depend on the full computeAll pipeline.
 */
function antiSquatAtAngle(
  theta_sa_rad: number,
  X_sp: number, H_sp: number,
  wheelbase: number,
  Y_cg: number,
  chain: ChainParams,
): number {
  const m1 = Math.tan(theta_sa_rad);
  const m2 = Math.tan(chain.chainForceAngle * DEG2RAD);
  if (Math.abs(m1 - m2) < 1e-9) return NaN; // parallel lines — no IC

  const x1 = X_sp, y1 = H_sp;
  const x2 = X_sp + chain.sprocketCenterX;
  const y2 = H_sp + chain.sprocketCenterY;

  const IC_x = (y2 - y1 + m1 * x1 - m2 * x2) / (m1 - m2);
  const IC_y = H_sp + m1 * (IC_x - X_sp);

  const denom = wheelbase - IC_x;
  if (Math.abs(denom) < 1e-9) return NaN;

  const slope_IC = (0 - IC_y) / denom;
  const h_front  = IC_y + slope_IC * (0 - IC_x);
  return (h_front / Y_cg) * 100;
}

// ─── Trail under fork dive ────────────────────────────────────────────────────

/**
 * Trail as a function of fork compression u_f (positive = dive).
 *
 * As the fork compresses, the steering axis tilts slightly forward,
 * reducing trail:
 *   T(u_f) = T_static − u_f × sin(α) × tan(α)
 *
 * Wheelbase also increases: W(u_f) = W_static + u_f × cos(α)
 * (not returned here but used for reference in sweep chart).
 *
 * Foale Ch. 2 derived form.
 */
function trailAtForkDive(
  u_f_mm: number,
  R_f: number,
  alpha_rad: number,
  forkOffset: number,
): number {
  const T_static = (R_f * Math.sin(alpha_rad) - forkOffset) / Math.cos(alpha_rad);
  return T_static - u_f_mm * Math.sin(alpha_rad) * Math.tan(alpha_rad);
}

// ─── Motion ratio (central difference) ───────────────────────────────────────

function motionRatio(
  u_mm: number,
  du: number,
  shockLenFn: (u: number) => number,
): number {
  const L_lo = shockLenFn(u_mm - du);
  const L_hi = shockLenFn(u_mm + du);
  // MR = du_wheel / du_shock; shock compresses as wheel rises → signs cancel
  // d(shock_length)/d(travel) is negative (shock gets shorter as wheel rises)
  const dL_du = (L_hi - L_lo) / (2 * du);
  // MR = |wheel travel change / shock compression change| = 1 / |dL/du|
  return Math.abs(1 / dL_du);
}

// ─── Main sweep function ──────────────────────────────────────────────────────

/**
 * Compute the full suspension travel sweep.
 *
 * @param gp         GeometryParams
 * @param susp       SuspensionParams (spring rate, shock travel)
 * @param chain      ChainParams (for anti-squat)
 * @param sweep      SweepParams (shock mount geometry)
 * @param Y_cg_mm    CoG height (mm) — from CoG module
 * @param du_mm      Step size (default 1 mm)
 * @returns SweepResults
 */
export function computeSweep(
  gp: GeometryParams,
  susp: SuspensionParams,
  chain: ChainParams,
  sweep: SweepParams,
  Y_cg_mm: number,
  du_mm = 1,
): SweepResults {
  const alpha_rad = gp.headAngle * DEG2RAD;
  const R_f = gp.frontWheelDia / 2;
  const k   = susp.springRateRear; // N/mm
  const maxTravel = susp.shockTravel / (sweep.linkageType === 'direct'
    ? 1.0
    : 1.0); // will be refined by actual MR; use shock travel as proxy for now

  // Travel range: 0 (static) to full wheel bump travel
  // We use shock travel * typical MR as proxy for wheel travel
  // Users can override via shockTravel input
  const nSteps = Math.round(maxTravel / du_mm) + 1;
  const points: SweepPoint[] = [];

  let theta_r0 = sweep.fourBar ? sweep.fourBar.rockerAngleStatic * DEG2RAD : 0;
  // Compute shock length at u=0 for compression reference
  const shockLen0 = sweep.linkageType === 'direct'
    ? shockLengthDirect(0, gp, sweep)
    : (() => { const r = shockLengthFourBar(0, gp, sweep, theta_r0); theta_r0 = r.theta_r; return r.shockLen; })();

  const shockLenFn = (u: number): number => {
    if (sweep.linkageType === 'direct') return shockLengthDirect(u, gp, sweep);
    return shockLengthFourBar(u, gp, sweep, theta_r0).shockLen;
  };

  for (let i = 0; i < nSteps; i++) {
    const u = i * du_mm;

    const theta_sa = swingarmAngleAtTravel(
      u, gp.rearAxleHeight, gp.swingarmPivotHeight, gp.swingarmLength,
    );

    let shockLen: number;
    if (sweep.linkageType === 'direct') {
      shockLen = shockLengthDirect(u, gp, sweep);
    } else {
      const res = shockLengthFourBar(u, gp, sweep, theta_r0);
      shockLen = res.shockLen;
      theta_r0 = res.theta_r; // warm-start for next step
    }

    const shockCompression = shockLen0 - shockLen;

    // Motion ratio (central difference; skip endpoints)
    let MR: number;
    if (u < du_mm || u > (nSteps - 2) * du_mm) {
      // forward/backward difference at endpoints
      const du2 = du_mm;
      const L1 = shockLenFn(u);
      const L2 = shockLenFn(u + (u < du_mm ? du2 : -du2));
      const sign = u < du_mm ? 1 : -1;
      const dL = sign * (L2 - L1) / du2;
      MR = Math.abs(1 / dL);
    } else {
      MR = motionRatio(u, du_mm, shockLenFn);
    }

    const WR = k * MR * MR;

    const AS = antiSquatAtAngle(
      theta_sa, gp.swingarmPivotX, gp.swingarmPivotHeight,
      gp.wheelbase, Y_cg_mm, chain,
    );

    // Trail: assume fork dives proportionally to rear bump (in-phase)
    // For pure rear sweep, trail stays static. For fork-dive sweep,
    // u is used as fork compression too. User can inspect independently.
    const trail = trailAtForkDive(u, R_f, alpha_rad, gp.forkOffset);

    points.push({
      travel_mm: u,
      swingarmAngleDeg: theta_sa * RAD2DEG,
      shockLength_mm: shockLen,
      shockCompression_mm: shockCompression,
      motionRatio: MR,
      wheelRate_Nmm: WR,
      antiSquatPct: isNaN(AS) ? 0 : AS,
      trail_mm: trail,
    });
  }

  return { points, static: points[0] };
}

File: chassis_sim/sweep.py
"""
chassis_sim/sweep.py — Suspension Travel Sweep Engine

Computes MR(u), WR(u), AS%(u), trail(u) over the full rear suspension
travel range by simulating swingarm kinematics at each 1 mm step.

Two rear suspension topologies:

  'direct'  — rear shock connects swingarm directly to frame (monoshock).
              Most modern sport bikes.

  'fourbar' — Pro-Link / Uni-Trak: rocker arm + pushrod interposed
              between swingarm and shock.  Loop closure solved by NR.

Reference:
  Foale (2006) Ch. 6, 11 — Motion ratio, wheel rate, anti-squat sweep
  Cossalter (2006) Ch. 5 — Rear suspension kinematics
"""

from __future__ import annotations
import math
import numpy as np
from dataclasses import dataclass, field
from typing import List, Literal, Optional

from .geometry import (
    BikeGeometry, ChainGeometry,
    compute_trail,
    compute_swingarm_angle_rad,
    compute_instant_centre,
    compute_anti_squat_pct,
    compute_anti_squat_swingarm_only,
)

DEG2RAD = math.pi / 180

# ── Shock mount configuration ─────────────────────────────────────────────────

@dataclass
class ShockMount:
    """
    Rear shock mount geometry.

    For 'direct' linkage:
      shock_arm_length, shock_arm_angle define the swingarm attachment.
      shock_top_x, shock_top_y define the frame attachment.

    For 'fourbar' linkage:
      The above define the pushrod attachment on the swingarm (S).
      Rocker pivot R, rocker length, pushrod length complete the 4-bar.
      Shock connects from rocker tip Q to shock_top.
    """
    linkage_type: Literal['direct', 'fourbar'] = 'direct'

    # Swingarm attachment
    shock_arm_length_mm: float = 120.0   # distance from pivot to attachment
    shock_arm_angle_deg: float = 85.0    # angle of arm relative to swingarm axis

    # Shock top mount (frame-fixed)
    shock_top_x_mm: float = 750.0
    shock_top_y_mm: float = 450.0

    # 4-bar linkage (only for linkage_type='fourbar')
    rocker_pivot_x_mm: float = 0.0
    rocker_pivot_y_mm: float = 0.0
    rocker_length_mm: float = 80.0
    pushrod_length_mm: float = 200.0
    rocker_angle_static_deg: float = 45.0


# ── Internal geometry helpers ─────────────────────────────────────────────────

def _swingarm_angle_at_travel(
    u_mm: float,
    H_ra_static: float,
    H_sp: float,
    L_sa: float,
) -> float:
    """
    Swingarm angle at wheel travel u (positive = compression/bump).

    H_ra(u) = H_ra_static + u  (axle rises with wheel travel)
    θ_sa(u) = arcsin((H_ra(u) − H_sp) / L_sa)

    The rear axle traces a circular arc of radius L_sa about the pivot.
    """
    H_ra_u = H_ra_static + u_mm
    sin_theta = (H_ra_u - H_sp) / L_sa
    sin_theta = max(-1.0, min(1.0, sin_theta))  # clamp for geometry limits
    return math.asin(sin_theta)


def _shock_attach_world(
    X_sp: float, H_sp: float,
    theta_sa_rad: float,
    arm_len: float, arm_angle_rad: float,
) -> tuple[float, float]:
    """World-frame position of shock attachment on swingarm."""
    angle = theta_sa_rad + arm_angle_rad
    return X_sp + arm_len * math.cos(angle), H_sp + arm_len * math.sin(angle)


def _dist(ax, ay, bx, by) -> float:
    return math.sqrt((bx - ax)**2 + (by - ay)**2)


# ── Direct monoshock ──────────────────────────────────────────────────────────

def _shock_length_direct(
    u_mm: float,
    geom: BikeGeometry,
    mount: ShockMount,
) -> float:
    theta = _swingarm_angle_at_travel(
        u_mm, geom.rear_axle_height_mm, geom.swingarm_pivot_height_mm, geom.swingarm_length_mm,
    )
    ax, ay = _shock_attach_world(
        geom.swingarm_pivot_x_mm, geom.swingarm_pivot_height_mm,
        theta, mount.shock_arm_length_mm, mount.shock_arm_angle_deg * DEG2RAD,
    )
    return _dist(ax, ay, mount.shock_top_x_mm, mount.shock_top_y_mm)


# ── 4-bar linkage (Pro-Link / Uni-Trak) ──────────────────────────────────────

def _solve_rocker_angle(
    S: tuple[float, float],
    R: tuple[float, float],
    L_rock: float,
    L_push: float,
    theta_r0: float,
) -> float:
    """
    Newton-Raphson solver for rocker angle θ_r.

    Loop closure: |Q − S|² = L_push²
    where Q = R + L_rock·[cos θ_r, sin θ_r]

    f(θ_r)  = (R_x + L_rock·cos θ_r − S_x)² + (R_y + L_rock·sin θ_r − S_y)² − L_push²
    f'(θ_r) = 2·[−L_rock·sin θ_r·(Q_x − S_x) + L_rock·cos θ_r·(Q_y − S_y)]
    """
    theta = theta_r0
    for _ in range(50):
        Qx = R[0] + L_rock * math.cos(theta)
        Qy = R[1] + L_rock * math.sin(theta)
        dx = Qx - S[0];  dy = Qy - S[1]
        f  = dx**2 + dy**2 - L_push**2
        df = 2 * (-L_rock * math.sin(theta) * dx + L_rock * math.cos(theta) * dy)
        if abs(df) < 1e-12:
            break
        dtheta = -f / df
        theta += dtheta
        if abs(dtheta) < 1e-9:
            break
    return theta


def _shock_length_fourbar(
    u_mm: float,
    geom: BikeGeometry,
    mount: ShockMount,
    theta_r0: float,
) -> tuple[float, float]:
    """Returns (shock_length, updated_rocker_angle)."""
    theta_sa = _swingarm_angle_at_travel(
        u_mm, geom.rear_axle_height_mm, geom.swingarm_pivot_height_mm, geom.swingarm_length_mm,
    )
    Sx, Sy = _shock_attach_world(
        geom.swingarm_pivot_x_mm, geom.swingarm_pivot_height_mm,
        theta_sa, mount.shock_arm_length_mm, mount.shock_arm_angle_deg * DEG2RAD,
    )
    R = (mount.rocker_pivot_x_mm, mount.rocker_pivot_y_mm)
    theta_r = _solve_rocker_angle(
        (Sx, Sy), R, mount.rocker_length_mm, mount.pushrod_length_mm, theta_r0,
    )
    Qx = R[0] + mount.rocker_length_mm * math.cos(theta_r)
    Qy = R[1] + mount.rocker_length_mm * math.sin(theta_r)
    shock_len = _dist(Qx, Qy, mount.shock_top_x_mm, mount.shock_top_y_mm)
    return shock_len, theta_r


# ── Trail under fork dive ─────────────────────────────────────────────────────

def _trail_at_fork_dive(
    u_f_mm: float,
    geom: BikeGeometry,
) -> float:
    """
    Trail as fork compresses by u_f mm.

    T(u_f) = T_static − u_f · sin α · tan α     ... Foale Ch. 2 derived

    As fork dives, steering axis tilts slightly forward → trail decreases.
    Braking dive reduces trail → lighter steering feedback while stopping.
    """
    T_static = compute_trail(geom)
    a = geom.alpha_rad
    return T_static - u_f_mm * math.sin(a) * math.tan(a)


# ── Sweep result dataclass ────────────────────────────────────────────────────

@dataclass
class SweepPoint:
    travel_mm: float
    swingarm_angle_deg: float
    shock_length_mm: float
    shock_compression_mm: float
    motion_ratio: float
    wheel_rate_Nmm: float
    anti_squat_pct: float
    anti_squat_swingarm_only_pct: float
    chain_contribution_pct: float
    trail_mm: float


@dataclass
class SweepResults:
    points: List[SweepPoint]
    static_point: SweepPoint  # u = 0

    def as_arrays(self) -> dict:
        """Return all channels as numpy arrays for plotting."""
        return {
            'travel_mm':           np.array([p.travel_mm for p in self.points]),
            'swingarm_angle_deg':  np.array([p.swingarm_angle_deg for p in self.points]),
            'motion_ratio':        np.array([p.motion_ratio for p in self.points]),
            'wheel_rate_Nmm':      np.array([p.wheel_rate_Nmm for p in self.points]),
            'anti_squat_pct':      np.array([p.anti_squat_pct for p in self.points]),
            'trail_mm':            np.array([p.trail_mm for p in self.points]),
        }

    def is_rising_rate(self, tol: float = 0.005) -> bool:
        """Returns True if MR monotonically decreases (rising wheel rate)."""
        mrs = [p.motion_ratio for p in self.points]
        diffs = [mrs[i+1] - mrs[i] for i in range(len(mrs)-1)]
        return all(d <= tol for d in diffs)

    def mr_range(self) -> tuple[float, float]:
        mrs = [p.motion_ratio for p in self.points]
        return min(mrs), max(mrs)

    def wr_range(self) -> tuple[float, float]:
        wrs = [p.wheel_rate_Nmm for p in self.points]
        return min(wrs), max(wrs)


# ── Main sweep function ───────────────────────────────────────────────────────

def compute_sweep(
    geom: BikeGeometry,
    spring_rate_Nmm: float,
    chain: ChainGeometry,
    mount: ShockMount,
    Y_cg_mm: float,
    wheel_travel_mm: float,
    du_mm: float = 1.0,
) -> SweepResults:
    """
    Compute suspension travel sweep.

    Parameters
    ----------
    geom             : BikeGeometry — static geometry
    spring_rate_Nmm  : rear coil spring rate (N/mm)
    chain            : ChainGeometry — for anti-squat computation
    mount            : ShockMount — shock attachment geometry
    Y_cg_mm          : CoG height (mm) — from compute_cog()
    wheel_travel_mm  : total wheel bump travel range (mm)
    du_mm            : step size (default 1 mm)

    Returns
    -------
    SweepResults with one SweepPoint per travel step.
    """
    steps = int(wheel_travel_mm / du_mm) + 1
    points: List[SweepPoint] = []

    # Warm-start rocker angle
    theta_r0 = mount.rocker_angle_static_deg * DEG2RAD if mount.linkage_type == 'fourbar' else 0.0

    # Reference shock length at u=0 (static position)
    if mount.linkage_type == 'direct':
        shock_len_0 = _shock_length_direct(0.0, geom, mount)
    else:
        shock_len_0, theta_r0 = _shock_length_fourbar(0.0, geom, mount, theta_r0)

    def shock_len_fn(u: float) -> float:
        nonlocal theta_r0
        if mount.linkage_type == 'direct':
            return _shock_length_direct(u, geom, mount)
        sl, tr = _shock_length_fourbar(u, geom, mount, theta_r0)
        return sl  # Note: theta_r0 NOT updated inside this closure (only in main loop)

    for i in range(steps):
        u = i * du_mm

        # Swingarm angle at this travel
        theta_sa = _swingarm_angle_at_travel(
            u, geom.rear_axle_height_mm,
            geom.swingarm_pivot_height_mm, geom.swingarm_length_mm,
        )

        # Shock length
        if mount.linkage_type == 'direct':
            shock_len = _shock_length_direct(u, geom, mount)
        else:
            shock_len, theta_r0 = _shock_length_fourbar(u, geom, mount, theta_r0)

        shock_compression = shock_len_0 - shock_len

        # Motion ratio — central difference (forward/backward at endpoints)
        if i == 0:
            L1 = shock_len
            L2 = shock_len_fn(u + du_mm)
            dL_du = (L2 - L1) / du_mm
        elif i == steps - 1:
            L1 = shock_len_fn(u - du_mm)
            L2 = shock_len
            dL_du = (L2 - L1) / du_mm
        else:
            L_lo = shock_len_fn(u - du_mm)
            L_hi = shock_len_fn(u + du_mm)
            dL_du = (L_hi - L_lo) / (2 * du_mm)

        # MR = |d(wheel_travel)/d(shock_compression)| = 1/|dL/du|
        MR = abs(1.0 / dL_du) if abs(dL_du) > 1e-10 else float('nan')
        WR = spring_rate_Nmm * MR**2 if not math.isnan(MR) else float('nan')

        # Anti-squat at this travel position (swingarm angle changes!)
        try:
            IC_x, IC_y = compute_instant_centre(geom, chain, theta_sa)
            AS_pct  = compute_anti_squat_pct(IC_x, IC_y, Y_cg_mm, geom.wheelbase_mm)
            AS_sa   = compute_anti_squat_swingarm_only(theta_sa, geom.wheelbase_mm, Y_cg_mm)
            chain_c = AS_pct - AS_sa
        except ValueError:
            AS_pct = float('nan')
            AS_sa  = float('nan')
            chain_c = float('nan')

        # Trail at equivalent fork dive
        trail = _trail_at_fork_dive(u, geom)

        points.append(SweepPoint(
            travel_mm=u,
            swingarm_angle_deg=math.degrees(theta_sa),
            shock_length_mm=shock_len,
            shock_compression_mm=shock_compression,
            motion_ratio=MR,
            wheel_rate_Nmm=WR,
            anti_squat_pct=AS_pct,
            anti_squat_swingarm_only_pct=AS_sa,
            chain_contribution_pct=chain_c,
            trail_mm=trail,
        ))

    return SweepResults(points=points, static_point=points[0])
